### ISSUE TITLE
Pilot should use struct to generate SDS config

### ISSUE THREAD

**Issue by JimmyCYJ** [istio/istio#16974]
(NOTE: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities/>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>
)

**Bug description**
Today, Pilot uses type Any in to generate SDS config. Although there is a mitigation to avoid Envoy generate multiple SDS clients, we still find that Envoy generate multiple SDS clients. The fix is to replace type Any with type struct in Pilot when generating SDS config.
**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**

**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version`)**

**How was Istio installed?**

**Environment where bug was observed (cloud vendor, OS, etc)**

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by howardjohn** [comment]
Can you add some details why Any is the cause of this and what the impact is?

**Comment by JimmyCYJ** [comment]
Thanks for asking!

Here are related issues, envoyproxy/envoy#5744, protocolbuffers/protobuf#5731.

Basically there is a call credential in SDS config, which has an Any type field. That field takes the marshalled result of file mount metadata config, and the marshalling for Any type is non deterministic. On receiving this SDS config, Envoy would generate different hash value for the registration of SDS client. Therefore, Envoy could create multiple SDS clients for identical SDS config.

Previously I add mitigation to make sure Pilot only creates one SDS config even with Any type, we introduce a map that stores the call credential config, and reuse that config instead of creating a new one.

Now we have more monitoring tools, which makes identifying this issue much easier. For example, there are dynamic SDS secrets in /config_dump endpoint, and a new debug endpoint at Citadel agent. We found that Envoy still creates multiple SDS connections for identical SDS config. These extra SDS pushes have performance impact on SDS push.

I have discussed with @lizan I am trying to fix this at Envoy side, so that we don't need to use Struct type for SDS config.
